megablocks/layers/glu_test.py

```diff
@@ -0,0 +1,156 @@
+
+    column_indices_t = row_indices.gather(0, gather_indices.long())
+    block_offsets_t = gather_indices.int()
+
+    zero = torch.zeros((1,), dtype=torch.int32, device=row_indices.device)
+    nnz_per_column = ops.histogram(column_indices, block_columns)
+    nnz_per_column = ops.inclusive_cumsum(nnz_per_column, 0)
+    offsets_t = torch.cat([zero, nnz_per_column])
+    return column_indices_t, offsets_t, block_offsets_t
+
+def _setup_topology(bs, sl, ffn_hidden_size, blocking=128, dtype=torch.bfloat16):
+    padded_tokens = bs * sl
+    padded_bins = torch.tensor([bs * sl]).type(torch.int32).cuda()
+    block_rows = padded_tokens // blocking
+    blocks_per_row = ffn_hidden_size // blocking
+
+    # equivalent to blocks_per_row in the 1 expert case
+    max_column_index = blocks_per_row
+
+    transpose_sort_end_bit = max(int(np.ceil(np.log2(max_column_index))), 1)
+
+    offsets = torch.arange(
+        0,
+        block_rows * blocks_per_row + 1,
+        blocks_per_row,
+        dtype=torch.int32,
+        device='cuda')
+
+    column_indices = ops.topology(
+            padded_bins, 
+            blocking,
+            block_rows,
+            blocks_per_row
+        )
+    data = torch.empty(
+        column_indices.numel(),
+        blocking,
+        blocking,
+        dtype=dtype,
+        device='meta') 
+
+    padded_tokens = bs * sl
+
+    shape = (
+        padded_tokens,
+        ffn_hidden_size
+    )
+
+    row_indices = stk.ops.row_indices(
+        shape, data, offsets, column_indices)
+    
+    column_indices_t, offsets_t, block_offsets_t = _sparse_transpose(
+        shape, row_indices, column_indices, offsets, blocking, transpose_sort_end_bit)
+    
+    return stk.Matrix(shape, data, row_indices, column_indices, offsets,
+                        column_indices_t, offsets_t, block_offsets_t)
+
+
+def test_modules(
+        hidden_size,
+        ffn_hidden_size,
+        grouped_mlp=False):
+    init_method = partial(torch.nn.init.normal_, mean=0.0, std=0.1)
+    args = Arguments(
+        hidden_size=hidden_size,
+        ffn_hidden_size=ffn_hidden_size,
+        moe_num_experts=1,
+        moe_top_k=1,
+        init_method=init_method,
+        memory_optimized_mlp=False,
+        mlp_type='glu',
+        grouped_mlp=grouped_mlp,
+        fp16=False,
+        bf16=True)
+
+    glu = testing.GLU(args)
+    dmoe_glu = GroupedGLU(args) if grouped_mlp else SparseGLU(args)
+
+    dmoe_glu.cuda(torch.cuda.current_device()).to(torch.bfloat16)
+    glu.cuda(torch.cuda.current_device()).to(torch.bfloat16)
+
+    with torch.no_grad():
+        glu.w1.copy_(dmoe_glu.w1.T)
+        glu.v1.copy_(dmoe_glu.v1.T)
+        glu.w2.copy_(dmoe_glu.w2)
+
+    return args, glu, dmoe_glu
+
+_DENSE_TESTS = (
+    (16, 1024, 512),
+    (8, 2048, 512),
+)
+
+class GLUTest(parameterized.TestCase):
+
+    @parameterized.parameters(*_DENSE_TESTS)
+    def testGLU_forward_grouped_mlp(self, bs, sl, hs):
```

Comment:
nit - after the underscore, the test name should also be CamelCase - i.e., testGLU_ForwardGroupedMLP